### DISCUSSION TITLE
Method "getHttpUrl" return wrong url for urls with ssh protocol with port

### DISCUSSION THREAD

**Discussion by websi** [renovatebot/renovate#28691] [Request Help] [answered]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab

### Please tell us more about your question or problem

In our setup we have private composer repositories an url with protocol ssh and define an non default port.

Example: `ssh://git@gitlab.example.com:22222/typo3-extensions/poll-pro.git`

The method `getHttpUrl` within `renovate/lib/util/git/url.ts` changed this to `https://gitlab.example.com:22222/typo3-extensions/poll-pro.git`.
The port `22222` is wrong for https.

Source code: https://github.com/renovatebot/renovate/blob/376a2a517b1f1bcc8f0ba4606c91b8d014d0ac02/lib/util/git/url.ts#L14-L16

In my local test the following line fixed the problem:

```diff
    const protocol = (0, regex_1.regEx)(/^https?$/).exec(parsedUrl.protocol)
        ? parsedUrl.protocol
        : 'https';
    parsedUrl.user = '';
    parsedUrl.token = token ?? '';
+   parsedUrl.port = '';
```

This is the very simple solution. Maybe the port can be configured via host rule.

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by rarkins** [comment]
It maybe makes sense to always strip the port from the transformed https url if the original was ssh? We know it won't be the same, so we may as well strip it?

**Reply by websi** [reply]
Yes the port is always not correct.

You also should think about to ignore the ssh to http tranformation at https://github.com/renovatebot/renovate/blob/376a2a517b1f1bcc8f0ba4606c91b8d014d0ac02/lib/modules/datasource/git-refs/base.ts#L45-L47
Because you also generate git config to rewrite ssh to https via the git feature "url.insteadOf" .
This is never used, because the url is rewritten in method "getRemoteUrlWithToken".

For us the "url.insteadOf" feature is the prefered one, because we can inject our own configuration via the runner environment.

---

I think i found an second issue. Within https://github.com/renovatebot/renovate/blob/376a2a517b1f1bcc8f0ba4606c91b8d014d0ac02/lib/util/git/auth.ts#L157 the configuration "customEnvVariables" is ignored. The method is called at https://github.com/renovatebot/renovate/blob/376a2a517b1f1bcc8f0ba4606c91b8d014d0ac02/lib/modules/datasource/git-refs/base.ts#L33

**Comment by rarkins** [comment, marked as answer]
https://github.com/renovatebot/renovate/pull/28722

**Reply by websi** [reply]
The new version fixed our problem. Thank you.
